src/solvers/strings/string_format_builtin_function.cpp

```diff
@@ -0,0 +1,482 @@
+  case format_specifiert::SCIENTIFIC_UPPER:
+  case format_specifiert::GENERAL_UPPER:
+  case format_specifiert::HEXADECIMAL_FLOAT_UPPER:
+  case format_specifiert::CHARACTER_UPPER:
+  case format_specifiert::DATE_TIME_UPPER:
+  case format_specifiert::BOOLEAN_UPPER:
+  case format_specifiert::STRING_UPPER:
+  case format_specifiert::HASHCODE_UPPER:
+  {
+    format_specifiert fs_lower = fs;
+    fs_lower.conversion = tolower(fs.conversion);
+    auto lower_case = eval_format_specifier(fs_lower, arg);
+    return make_range(lower_case).map([](const mp_integer &c) {
+      // TODO: incomplete
+      if('a' <= c && c <= 'z')
+        return c + 0x20;
+      return c;
+    });
+  }
+  case format_specifiert::OCTAL_INTEGER:
+    /// \todo Conversion of octal is not implemented.
+  case format_specifiert::GENERAL:
+    /// \todo Conversion for format specifier general is not implemented.
+  case format_specifiert::HEXADECIMAL_FLOAT:
+    /// \todo Conversion of hexadecimal float is not implemented.
+  case format_specifiert::DATE_TIME:
+    /// \todo Conversion of date-time is not implemented
+    UNIMPLEMENTED;
+  }
+
+  INVARIANT(false, "format specifier must belong to [bBhHsScCdoxXeEfgGaAtT%n]");
+}
+
+optionalt<exprt> string_format_builtin_functiont::eval(
+  const std::function<exprt(const exprt &)> &get_value) const
+{
+  if(!format_string.has_value())
+    return {};
+
+  const std::vector<format_elementt> format_strings =
+    parse_format_string(format_string.value());
+  std::vector<mp_integer> result_vector;
+  std::size_t arg_count = 0;
+
+  for(const format_elementt &fe : format_strings)
+  {
+    if(fe.is_format_specifier())
+    {
+      const format_specifiert &fs = fe.get_format_specifier();
+      if(
+        fs.conversion != format_specifiert::PERCENT_SIGN &&
+        fs.conversion != format_specifiert::LINE_SEPARATOR)
+      {
+        std::vector<mp_integer> evaluated_char_vector;
+
+        if(fs.index == -1)
+        {
+          INVARIANT(
+            arg_count < inputs.size(),
+            "number of format must match specifiers");
+          if(auto arg_value = eval_string(inputs[arg_count++], get_value))
+            evaluated_char_vector = eval_format_specifier(fs, *arg_value);
+          else
+            return {};
+        }
+        else
+        {
+          INVARIANT(fs.index > 0, "index in format should be positive");
+          INVARIANT(
+            static_cast<std::size_t>(fs.index) <= inputs.size(),
+            "number of format must match specifiers");
+
+          // first argument `args[0]` corresponds to index 1
+          if(auto arg_value = eval_string(inputs[fs.index - 1], get_value))
+            evaluated_char_vector = eval_format_specifier(fs, *arg_value);
+          else
+            return {};
+        }
+        std::move(
+          evaluated_char_vector.begin(),
+          evaluated_char_vector.end(),
+          std::back_inserter(result_vector));
+      }
+      // TODO: handle this case
```

Comment:
adding the `%%` and `%n` cases should be relatively easy

Reply:
Doing now as follow-up PR.

Reply:
Are there tickets that track all these todos?